### PR TITLE
Lewis

### PR DESCRIPTION
**Pull request by lew101**
Just got to add mapping then I'll be done.

I have added a launch file to simplify things but I can't figure ou how to localise in rViz programmatically so I can't start everything from the launch file. Some advice would be great regarding that.

### CODE REVIEW

**Comment by akhmaddias** [comment]
> I have added a launch file to simplify things but I can't figure ou how to localise in rViz programmatically so I can't start everything from the launch file. Some advice would be great regarding that.

I was also trying to do so but with no success.

Other than that your solution is great.

File: minitask4/Lewis/scripts/mapping.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+    grid_x = floor((world_x - ORIGIN_X) / RESOLUTION)
+    grid_y = floor((world_y - ORIGIN_Y) / RESOLUTION)
+
+    if (grid_x >= SIZE_X or grid_y >= SIZE_Y):
+        return None  # If the coordinates are after the end of the grid
+
+    return (grid_x, grid_y)
+
+def to_world(grid_x, grid_y):
```
Comment (akhmaddias) [inline minitask4/Lewis/scripts/mapping.py:110]: I believe this function is not used anywhere

**Review by akhmaddias — APPROVED** [review]
LGTM!

File: minitask4/Lewis/scripts/mapping.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        self.initialise_grid(UNVISITED)
+        self.setup_plot()
+
+    def odom_callback(self, odom):
+        '''
+        Callback method for the odometry subscriber.
+        '''
+        world_x, world_y = odom.pose.pose.position.x, odom.pose.pose.position.y
+
+        (grid_x, grid_y) = to_grid(world_x, world_y)
```
Comment (bhenaghan) [inline minitask4/Lewis/scripts/mapping.py:45]: You don't need the brackets here. This is more of a style thing as they might make it clearer, however I think it should be consistent with the line above (`43`).

File: minitask4/Lewis/scripts/mapping.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+def to_grid(world_x, world_y):
+    '''
+    Will return occupancy grid coordinates given real-world coordinates
+    or None if the coordinates fall outside the range of the occupancy grid.
+    '''
+    if (world_x < ORIGIN_X or world_y < ORIGIN_Y):
+        return None  # If the coordinates are before the start of the grid
+
+    grid_x = floor((world_x - ORIGIN_X) / RESOLUTION)
```
Comment (bhenaghan) [inline minitask4/Lewis/scripts/mapping.py:102]: You can replace the `floor` with integer division (`//`)

File: minitask4/Lewis/scripts/mapping.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    grid_y = floor((world_y - ORIGIN_Y) / RESOLUTION)
+
+    if (grid_x >= SIZE_X or grid_y >= SIZE_Y):
+        return None  # If the coordinates are after the end of the grid
```
Comment (bhenaghan) [inline minitask4/Lewis/scripts/mapping.py:106]: This is going to cause an issue when the other code assumes it should receive a tuple. I think its clearer to raise an exception here.

File: minitask4/Lewis/scripts/mapping.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+    return (grid_x, grid_y)
```
Comment (bhenaghan) [inline minitask4/Lewis/scripts/mapping.py:108]: again, the brackets are not strictly needed but its a style/preference thing - lets just make sure its consistent

File: minitask4/Lewis/scripts/mapping.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+    return (world_x, world_y)
+
+def to_cell_index(grid_x, grid_y):
+    '''
+    Will return an array index given a grid coordinates.
+    This is mirrored in X and Y so that the map accurately represents
+    the path that the bot has followed.
+    '''
+    return int((SIZE_X * SIZE_Y) - (grid_x * SIZE_X + grid_y))
```
Comment (bhenaghan) [inline minitask4/Lewis/scripts/mapping.py:132]: The `int` case here is redundant as long as grid_x and grid_y are integers

**Review by BenH14 — APPROVED** [review]
Looks really good!

File: minitask4/Lewis/scripts/mapping.py (hunk 3 above)
Comment (bhenaghan) [reply minitask4/Lewis/scripts/mapping.py:102]: Oops this isn't valid because we're using python2.

File: minitask4/Lewis/scripts/mapping.py (hunk 1 above)
Comment (lew101) [reply minitask4/Lewis/scripts/mapping.py:110]: It isn't but Ayse wanted a function that could convert "grid coordinates to world coordinates considering the origin, grid size and resolution"

**Pull request merged**